File: packages/go/hushspec/resolve.go
package hushspec

import (
	"fmt"
	"os"
	"path/filepath"
	"slices"
	"strings"
)

// LoadedSpec pairs a parsed HushSpec with the canonical path it was loaded from.
type LoadedSpec struct {
	Source string
	Spec   *HushSpec
}

// ResolveLoader loads a HushSpec referenced by an extends field.
// reference is the extends value; from is the source of the referencing document.
type ResolveLoader func(reference string, from string) (*LoadedSpec, error)

// Resolve flattens the extends chain of a parsed HushSpec by repeatedly loading
// and merging parent documents via the provided loader.
func Resolve(spec *HushSpec, source string, loader ResolveLoader) (*HushSpec, error) {
	if loader == nil {
		loader = loadFromFilesystem
	}

	stack := make([]string, 0, 4)
	if source != "" {
		stack = append(stack, source)
	}
	return resolveInner(spec, source, loader, stack)
}

// ResolveFile loads a HushSpec from disk and flattens its extends chain.
func ResolveFile(path string) (*HushSpec, error) {
	source, err := filepath.Abs(path)
	if err != nil {
		return nil, fmt.Errorf("failed to resolve path %q: %w", path, err)
	}
	source, err = filepath.EvalSymlinks(source)
	if err != nil {
		return nil, fmt.Errorf("failed to canonicalize %q: %w", source, err)
	}
	content, err := os.ReadFile(source)
	if err != nil {
		return nil, fmt.Errorf("failed to read HushSpec at %s: %w", source, err)
	}
	spec, err := Parse(string(content))
	if err != nil {
		return nil, fmt.Errorf("failed to parse HushSpec at %s: %w", source, err)
	}
	return Resolve(spec, source, loadFromFilesystem)
}

func resolveInner(spec *HushSpec, source string, loader ResolveLoader, stack []string) (*HushSpec, error) {
	if spec == nil || spec.Extends == "" {
		return spec, nil
	}

	loaded, err := loader(spec.Extends, source)
	if err != nil {
		return nil, err
	}

	for index, entry := range stack {
		if entry == loaded.Source {
			cycle := append(slices.Clone(stack[index:]), loaded.Source)
			return nil, fmt.Errorf("circular extends detected: %s", joinChain(cycle))
		}
	}

	nextStack := append(stack, loaded.Source)
	parent, err := resolveInner(loaded.Spec, loaded.Source, loader, nextStack)
	if err != nil {
		return nil, err
	}

	return Merge(parent, spec), nil
}

func loadFromFilesystem(reference string, from string) (*LoadedSpec, error) {
	resolvedPath := reference
	if !filepath.IsAbs(reference) && from != "" {
		resolvedPath = filepath.Join(filepath.Dir(from), reference)
	}

	absPath, err := filepath.Abs(resolvedPath)
	if err != nil {
		return nil, fmt.Errorf("failed to resolve path %q: %w", resolvedPath, err)
	}
	canonical, err := filepath.EvalSymlinks(absPath)
	if err != nil {
		return nil, fmt.Errorf("failed to canonicalize %q: %w", absPath, err)
	}
	content, err := os.ReadFile(canonical)
	if err != nil {
		return nil, fmt.Errorf("failed to read HushSpec at %s: %w", canonical, err)
	}
	spec, err := Parse(string(content))
	if err != nil {
		return nil, fmt.Errorf("failed to parse HushSpec at %s: %w", canonical, err)
	}
	return &LoadedSpec{Source: canonical, Spec: spec}, nil
}

func joinChain(chain []string) string {
	return strings.Join(chain, " -> ")
}

File: packages/python/hushspec/resolve.py
from __future__ import annotations

from dataclasses import dataclass
from pathlib import Path
from typing import Callable

from hushspec.merge import merge
from hushspec.parse import parse
from hushspec.schema import HushSpec


@dataclass
class LoadedSpec:
    source: str
    spec: HushSpec


Resolver = Callable[[str, str | None], LoadedSpec]


def resolve(
    spec: HushSpec,
    *,
    source: str | None = None,
    loader: Resolver | None = None,
) -> tuple[bool, HushSpec | str]:
    stack = [source] if source is not None else []
    return _resolve_inner(spec, source, loader or _load_from_filesystem, stack)


def resolve_or_raise(
    spec: HushSpec,
    *,
    source: str | None = None,
    loader: Resolver | None = None,
) -> HushSpec:
    ok, result = resolve(spec, source=source, loader=loader)
    if not ok:
        raise ValueError(result)
    return result


def resolve_file(path: str | Path) -> tuple[bool, HushSpec | str]:
    source = str(Path(path).resolve())
    try:
        content = Path(source).read_text()
    except OSError as exc:
        return False, f"failed to read HushSpec at {source}: {exc}"
    ok, parsed = parse(content)
    if not ok:
        return False, f"failed to parse HushSpec at {source}: {parsed}"
    return resolve(parsed, source=source, loader=_load_from_filesystem)


def _resolve_inner(
    spec: HushSpec,
    source: str | None,
    loader: Resolver,
    stack: list[str],
) -> tuple[bool, HushSpec | str]:
    if spec.extends is None:
        return True, spec

    try:
        loaded = loader(spec.extends, source)
    except Exception as exc:  # pragma: no cover - exercised through public API
        return False, str(exc)

    if loaded.source in stack:
        cycle = stack[stack.index(loaded.source) :] + [loaded.source]
        return False, f"circular extends detected: {' -> '.join(cycle)}"

    stack.append(loaded.source)
    ok, parent = _resolve_inner(loaded.spec, loaded.source, loader, stack)
    stack.pop()
    if not ok:
        return False, parent

    return True, merge(parent, spec)


def _load_from_filesystem(reference: str, source: str | None) -> LoadedSpec:
    path = Path(reference)
    if not path.is_absolute():
        path = Path(source).parent / path if source is not None else path.resolve()
    canonical = path.resolve()
    content = canonical.read_text()
    ok, parsed = parse(content)
    if not ok:
        raise ValueError(f"failed to parse HushSpec at {canonical}: {parsed}")
    return LoadedSpec(source=str(canonical), spec=parsed)
